// @datorama/utils/dist/function/index.js
import _debounce from 'lodash.debounce';
import _throttle from 'lodash.throttle';
import _curry from 'lodash.curry';
/**
 *
 * @param func
 * @param wait
 * @param options
 */
export function throttle(func, wait, options) {
    return _throttle(func, wait, options);
}
/**
 *
 * @param func
 * @param wait
 * @param options
 */
export function debounce(func, wait, options) {
    return _debounce(func, wait, options);
}
/**
 *
 * @param func
 * @param {number} arity
 */
export function curry(func, arity) {
    if (arity === void 0) { arity = func.length; }
    return _curry(func, arity);
}
/**
 *  Return a composed function
 * @param funcs
 * @returns {(value) => any}
 */
export function compose() {
    var funcs = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        funcs[_i] = arguments[_i];
    }
    return function (value) {
        return funcs.reduce(function (value, func) { return func(value); }, value);
    };
}
/**
 * Return the name of the function
 * @example getFunctionName(function myAwesomeFunc(){}) ---> 'myAwesomeFunc'
 * @param func
 */
export function getFunctionName(func) {
    if (func == null || typeof func !== 'function') {
        return '';
    }
    // check if we've the name property (ES6)
    if (func.hasOwnProperty('name')) {
        return func.name;
    }
    // ES5
    var funcName = func.toString();
    funcName = funcName.substr('function '.length);
    funcName = funcName.substr(0, funcName.indexOf('('));
    return funcName;
}
